lib/duffel_api/webhook_event.rb

```diff
@@ -0,0 +1,119 @@
+# frozen_string_literal: true
+
+require "base16"
+require "openssl"
+
+module DuffelAPI
+  module WebhookEvent
+    # An exception raised internally within the library - but not exposed - if the
+    # webhook signature provided does not match the format of a valid signature
+    class InvalidRequestSignatureError < StandardError; end
+
+    SIGNATURE_REGEXP = /\At=(.+),v1=(.+)\z/.freeze
+
+    SHA_256 = OpenSSL::Digest.new("sha256")
+
+    class << self
+      # Checks if a webhook event you received was a genuine webhook event from Duffel by
+      # checking that it was signed with your shared secret.
+      #
+      # Assuming that you've kept that secret secure and only shared it with Duffel,
+      # this can give you confidence that a webhook event was genuinely sent by Duffel.
+      #
+      # @param [String] request_body The raw body of the received request
+      # @param [String] request_signature The signature provided with the received
+      #   request, found in the `X-Duffel-Signature` request header
+      # @param [String] webhook_secret The secret of the webhook, registered with Duffel
+      # @return [Boolean] whether the webhook signature matches
+      def genuine?(request_body:, request_signature:, webhook_secret:)
+        parsed_signature = parse_signature!(request_signature)
+
+        calculated_hmac = calculate_hmac(
+          payload: request_body,
+          secret: webhook_secret,
+          timestamp: parsed_signature[:timestamp],
+        )
+
+        secure_compare(calculated_hmac, parsed_signature[:v1])
+      rescue InvalidRequestSignatureError
+        # If the signature doesn't even look like a valid one, then the webhook
+        # event can't be genuine
+        false
+      end
+
+      private
+
+      # Calculates the signature for a request body in the same way that the Duffel API
+      # does it
+      #
+      # @param [String] secret
+      # @param [String] payload
+      # @param [String] timestamp
+      # @return [Stringop]
+      def calculate_hmac(secret:, payload:, timestamp:)
+        signed_payload = timestamp + "." + payload
```

Comment:
I'd force these values to be strings via interpolation.
```suggestion
        signed_payload = %(#{timestamp}.#{payload})
```

Reply:
Good call! I've no idea why I did it this way honestly 😹